vault/token_store.go

```diff
@@ -2205,6 +2214,87 @@ func (ts *TokenStore) handleCreateCommon(ctx context.Context, req *logical.Reque
 			logical.ErrInvalidRequest
 	}
 
+	// Verify the entity alias
+	overwriteEntityID := ""
+	if data.EntityAlias != "" {
+		// Parameter is only allowed in combination with token role
+		if role == nil {
+			return logical.ErrorResponse("'entity_alias' is only allowed in combination with token role"), logical.ErrInvalidRequest
+		}
+
+		// Check if provided entity alias name is in the allowed entity aliases list
+		match := false
+		for _, allowedAlias := range role.AllowedEntityAliases {
+			// Check if there is a match
+			if strings.EqualFold(allowedAlias, data.EntityAlias) {
+				match = true
+				break
+			}
+
+			// Only continue when this allowed alias is a glob pattern
+			if !strings.Contains(allowedAlias, "*") {
+				continue
+			}
+
+			// Split by asterisk to get prefix and suffix
+			split := strings.Split(allowedAlias, "*")
+
+			// Multiple asterisks are not allowed. Skip this invalid glob pattern.
+			if len(split) != 2 {
+				continue
+			}
+
+			// Check if prefix matches
+			if !strings.HasPrefix(data.EntityAlias, split[0]) {
+				continue
+			}
+
+			// Check if suffix matches
+			if !strings.HasSuffix(data.EntityAlias, split[1]) {
+				continue
+			}
+
+			// Found a match
+			match = true
+			break
+		}
+
+		// Throw an error if it does not match
+		if !match {
+			return logical.ErrorResponse("invalid 'entity_alias' value"), logical.ErrInvalidRequest
+		}
+
+		// Get mount accessor which is required to lookup entity alias
+		mountValidationResp := ts.core.router.MatchingMountByAccessor(req.MountAccessor)
+		if mountValidationResp == nil {
+			return logical.ErrorResponse("auth token mount accessor not found"), nil
+		}
+
+		// Create alias for later processing
+		alias := &logical.Alias{
+			Name:          data.EntityAlias,
+			MountAccessor: mountValidationResp.Accessor,
+			MountType:     mountValidationResp.Type,
+		}
+
+		// Create or fetch entity from entity alias
+		entity, err := ts.core.identityStore.CreateOrFetchEntity(ctx, alias)
+		if err != nil {
+			return logical.ErrorResponse(err.Error()), nil
```

Comment:
Am I right in thinking that if you return a `logical.ErrorResponse, nil`, Vault returns a 4xx response; whereas if you return `nil, err`, Vault returns a 5xx response? If that's the case, I kind of think that 4xx is more for if the _user_ did something that needs correction, and 5xx is more for like, something on our end failed or we're experiencing a possibly transient issue.

Anyways, I wonder if this might be better suited to the 5xx family because it's coming from an internal call rather than as an obvious instance of bad user input.

Reply:
Good catch! You are right. Thanks!